File: demo_client.cc
#include <iostream>
#include <memory>
#include <string>
#include <cstdlib>

#include "demo/v1.pb.h"
#include <grpcpp/grpcpp.h>
#include "demo/v1.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using demo::v1::DemoService;
using demo::v1::EchoRequest;
using demo::v1::EchoResponse;
using demo::v1::HealthResponse;
using google::protobuf::Empty;

int main() {
    const char* addr_env = std::getenv("SERVER_ADDR");
    std::string server_addr = addr_env ? addr_env : "localhost:50051";

    auto channel = grpc::CreateChannel(server_addr, grpc::InsecureChannelCredentials());
    auto stub = DemoService::NewStub(channel);

    // Echo
    {
        EchoRequest request;
        request.set_message("hello");
        EchoResponse response;
        ClientContext ctx;
        Status status = stub->Echo(&ctx, request, &response);
        if (!status.ok()) {
            std::cerr << "Echo RPC failed: " << status.error_message() << std::endl;
            return 1;
        }
        std::cout << "Echo response: " << response.message() << std::endl;
    }

    // Health
    {
        Empty request;
        HealthResponse response;
        ClientContext ctx;
        Status status = stub->Health(&ctx, request, &response);
        if (!status.ok()) {
            std::cerr << "Health RPC failed: " << status.error_message() << std::endl;
            return 1;
        }
        std::cout << "Health status: " << response.status() << std::endl;
    }

    return 0;
}

File: demo_client.rb
$LOAD_PATH.unshift File.join(__dir__, 'lib')
require 'grpc'
require 'demo/v1_services_pb'

def main
  server_addr = ENV.fetch('SERVER_ADDR', 'localhost:50051')
  stub = Demo::V1::DemoService::Stub.new(server_addr, :this_channel_is_insecure)

  echo_resp = stub.echo(Demo::V1::EchoRequest.new(message: 'hello'))
  puts "Echo response: #{echo_resp.message}"

  health_resp = stub.health(Google::Protobuf::Empty.new)
  puts "Health status: #{health_resp.status}"
end

main
